### PR TITLE
Additional DSignal functions

### PR DESCRIPTION
**Pull request by blaxill**
- delayN and delayI, delay a DSignal and leave the initial values after a
reset undefined. (Similar to Clash.Signal.delay)
- delayedFold, a tree fold by simple combinatorial application that
resolves its type level delay.

### CODE REVIEW

**Comment by blaxill** [comment]
This also fixes DSignal doctests

**Comment by blaxill** [comment]
I've rebased onto your merged fixes (specifically keeping your Clash.Explicit.Signal.Delayed)

**Comment by blaxill** [comment]
@christiaanb are there any other changes you'd like?

**Comment by martijnbastiaan** [comment]
@blaxill Could you add a test case in https://github.com/clash-lang/clash-compiler/blob/master/testsuite/Main.hs? I'm willing to do it if you haven't got the time. You can basically copy the approach here: https://github.com/clash-lang/clash-compiler/blob/master/tests/shouldwork/Vector/VFold.hs. Run your testcase with `cabal new-run testsuite -- -p VFold`. (Where `VFold` is replaced by your test case name of course.) Don't worry if you get some failures due to missing simulators: some are annoyingly difficult to install, and Travis will run them for you anyway.

**Comment by blaxill** [comment]
Sure, I think I'll have time in the next day or so.

**Comment by martijnbastiaan** [comment]
Could you also rebase this branch on `master`? We've switched CI providers in the meantime.

File: clash-prelude/src/Clash/Signal/Delayed.hs [hunk 1]
```diff
@@ -5,20 +5,31 @@ License    :  BSD2 (see the file LICENSE)
...
+{-# LANGUAGE CPP                  #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE KindSignatures       #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE UndecidableInstances #-}
+#if __GLASGOW_HASKELL__ >= 806
```
Comment (martijnbastiaan) [inline clash-prelude/src/Clash/Signal/Delayed.hs]: This is for GHC 8.0.6, right? We don't support <8.2, so I think you can remove this.

Comment (blaxill) [reply clash-prelude/src/Clash/Signal/Delayed.hs]: Actually kind of confusingly, 806 refers to [8.6.0 (see here)](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/intro.html#ghc-version-numbering-policy). This case is needed for 8.6.0 and above as GHC starts interpreting all unqualified cases of * as `Type` (interfering with type level multiplication). We could also enable NoStarIsType in the cabal config or use ugly qualifications, but it looks like Clash is using this same construct elsewhere, and a general change like that would be better in its own PR? ([8.6.0 StarIsType migration notes](https://ghc.haskell.org/trac/ghc/wiki/Migration/8.6#StarIsType))

Comment (martijnbastiaan) [reply clash-prelude/src/Clash/Signal/Delayed.hs]: I see, thanks for the pointer! And I agree, that change is better suited for another PR.

**Comment by martijnbastiaan** [comment]
Thank you @blaxill !

**Pull request merged**